lib/que/metrics.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+require "rack"
+require "benchmark"
+require "prometheus/client"
+require "prometheus/middleware/exporter"
+
+module Que
+  class Metrics
+    def initialize(labels: {})
+      @registry = Prometheus::Client::Registry.new
+
+      @jobs_worked_total = @registry.counter(:que_jobs_worked_total, "Counter for all jobs processed", labels)
+      @jobs_worked_seconds_total = @registry.counter(:que_jobs_worked_seconds_total, "Sum of the time taken processing each job class", labels)
+      @jobs_latency_seconds_total = @registry.counter(:que_jobs_latency_seconds_total, "Sum of time spent by job and priority waiting in queue", labels)
+      @worker_running_seconds_total = @registry.counter(:que_worker_running_seconds_total, "Time since starting to work jobs", labels)
+      @worker_job_acquire_total = @registry.counter(:que_worker_job_acquire_total, "Counter of number of attempts to acquire jobs", labels)
+      @worker_job_acquire_seconds_total = @registry.counter(:que_worker_job_acquire_seconds_total, "Sum of time taken to acquire jobs", labels)
+    end
+
+    def expose(port: 8080)
+      Rack::Handler::WEBrick.run(
+        Prometheus::Middleware::Exporter.new(Proc.new {}, registry: @registry), Port: port,
+      )
+    end
+
+    def start_work
+      stop = false
+
+      Thread.new do
+        last_timestamp = monotonic_now
+
+        until stop
+          timestamp = monotonic_now
+          @worker_running_seconds_total.increment({}, timestamp - last_timestamp)
+          last_timestamp = timestamp
+
+          sleep(1) # delay the next update for 1s
+        end
+      end
+
+      -> { stop = true }
+    end
+
+    def acquire_job(labels, &block)
```

Comment:
How would you feel about calling all these methods `trace_X` or `track_X` or something? I think it might make it easier to read when used in the worker (to indicate that the method doesn't actually, for example, acquire a job).

Reply:
Yeah I think this is a really good point.